File: ZezeJava/src/Zeze/Serialize/SerializeHelper.java
package Zeze.Serialize;

import Zeze.*;

public class SerializeHelper {
	public static <T> tangible.Func1Param<ByteBuffer, T> CreateDecodeFunc() {
		var type = T.class;

//C# TO JAVA CONVERTER WARNING: Unsigned integer types have no direct equivalent in Java:
//ORIGINAL LINE: if (type == typeof(byte))
		if (type == Byte.class) {
			return (tangible.Func1Param<ByteBuffer, T>)(Delegate)((ByteBuffer arg) -> bb.ReadByte());
		}

		if (type == Short.class) {
			return (tangible.Func1Param<ByteBuffer, T>)(Delegate)((ByteBuffer arg) -> bb.ReadShort());
		}

		if (type == Integer.class) {
			return (tangible.Func1Param<ByteBuffer, T>)(Delegate)((ByteBuffer arg) -> bb.ReadInt());
		}

		if (type == Long.class) {
			return (tangible.Func1Param<ByteBuffer, T>)(Delegate)((ByteBuffer arg) -> bb.ReadLong());
		}

		if (type == String.class) {
			return (tangible.Func1Param<ByteBuffer, T>)(Delegate)((ByteBuffer arg) -> bb.ReadString());
		}

		if (type == Zeze.Net.Binary.class) {
			return (tangible.Func1Param<ByteBuffer, T>)(Delegate)((ByteBuffer arg) -> bb.ReadBinary());
		}

		if (type.IsAssignableTo(Serializable.class)) {
			return (tangible.Func1Param<ByteBuffer, T>)Delegate.CreateDelegate(tangible.Func1Param<ByteBuffer, T>.class, SerializeHelper.class.GetMethod("CreateSerialiableDecodeFunc").MakeGenericMethod(type));
		}

		return null;
	}

//C# TO JAVA CONVERTER TODO TASK: The C# 'new()' constraint has no equivalent in Java:
//ORIGINAL LINE: public static T CreateSerialiableDecodeFunc<T>(ByteBuffer buf) where T : Serializable, new()
	public static <T extends Serializable> T CreateSerialiableDecodeFunc(ByteBuffer buf) {
		var value = new T();
		value.Decode(buf);
		return value;
	}

	public static <T> tangible.Action2Param<ByteBuffer, T> CreateEncodeFunc() {
		var type = T.class;

//C# TO JAVA CONVERTER WARNING: Unsigned integer types have no direct equivalent in Java:
//ORIGINAL LINE: if (type == typeof(byte))
		if (type == Byte.class) {
//C# TO JAVA CONVERTER WARNING: Unsigned integer types have no direct equivalent in Java:
//ORIGINAL LINE: return (Action<ByteBuffer, T>)(Delegate)(new Action<ByteBuffer, byte>((ByteBuffer buf, byte x) => buf.WriteByte(x)));
			return (tangible.Action2Param<ByteBuffer, T>)(Delegate)((ByteBuffer arg1, byte arg2) -> buf.WriteByte(x));
		}

		if (type == Short.class) {
			return (tangible.Action2Param<ByteBuffer, T>)(Delegate)((ByteBuffer arg1, short arg2) -> buf.WriteShort(x));
		}

		if (type == Integer.class) {
			return (tangible.Action2Param<ByteBuffer, T>)(Delegate)((ByteBuffer arg1, int arg2) -> buf.WriteInt(x));
		}

		if (type == Long.class) {
			return (tangible.Action2Param<ByteBuffer, T>)(Delegate)((ByteBuffer arg1, long arg2) -> buf.WriteLong(x));
		}

		if (type == String.class) {
			return (tangible.Action2Param<ByteBuffer, T>)(Delegate)((ByteBuffer arg1, String arg2) -> buf.WriteString(x));
		}

		if (type == Zeze.Net.Binary.class) {
			return (tangible.Action2Param<ByteBuffer, T>)(Delegate)((ByteBuffer arg1, Zeze.Net.Binary arg2) -> buf.WriteBinary(x));
		}

		if (T.class.IsSubclassOf(Serializable.class)) {
			return (ByteBuffer buf, T x) -> (x instanceof Serializable ? (Serializable)x : null).Encode(buf);
		}

		return null;
	}
}

File: Zeze/Serialize/SerializeHelper.cs
using System;
using Zeze.Net;

namespace Zeze.Serialize
{
    public class SerializeHelper
    {
        public static Func<ByteBuffer, T> CreateDecodeFunc<T>()
        {
            var type = typeof(T);

            if (type == typeof(bool))
            {
                return (Func<ByteBuffer, T>)(Delegate)(new Func<ByteBuffer, bool>(
                    (ByteBuffer bb) => bb.ReadBool()));
            }

            if (type == typeof(byte))
            {
                return (Func<ByteBuffer, T>)(Delegate)(new Func<ByteBuffer, byte>(
                    (ByteBuffer bb) => (byte)bb.ReadLong()));
            }

            if (type == typeof(short))
            {
                return (Func<ByteBuffer, T>)(Delegate)(new Func<ByteBuffer, short>(
                    (ByteBuffer bb) => (short)bb.ReadLong()));
            }

            if (type == typeof(int))
            {
                return (Func<ByteBuffer, T>)(Delegate)(new Func<ByteBuffer, int>(
                    (ByteBuffer bb) => bb.ReadInt()));
            }

            if (type == typeof(long))
            {
                return (Func<ByteBuffer, T>)(Delegate)(new Func<ByteBuffer, long>(
                    (ByteBuffer bb) => bb.ReadLong()));
            }

            if (type == typeof(string))
            {
                return (Func<ByteBuffer, T>)(Delegate)(new Func<ByteBuffer, string>(
                    (ByteBuffer bb) => bb.ReadString()));
            }

            if (type == typeof(double))
            {
                return (Func<ByteBuffer, T>)(Delegate)(new Func<ByteBuffer, double>(
                    (ByteBuffer bb) => bb.ReadDouble()));
            }

            if (type == typeof(float))
            {
                return (Func<ByteBuffer, T>)(Delegate)(new Func<ByteBuffer, float>(
                    (ByteBuffer bb) => bb.ReadFloat()));
            }

            if (type == typeof(Binary))
            {
                return (Func<ByteBuffer, T>)(Delegate)(new Func<ByteBuffer, Binary>(
                    (ByteBuffer bb) => bb.ReadBinary()));
            }

            if (type.IsAssignableTo(typeof(Serializable)))
            {
                return (Func<ByteBuffer, T>)Delegate.CreateDelegate(
                    typeof(Func<ByteBuffer, T>),
                    typeof(SerializeHelper)
                    .GetMethod(nameof(CreateSerialiableDecodeFunc))
                    .MakeGenericMethod(type));
            }

            return null;
        }

        public static T CreateSerialiableDecodeFunc<T>(ByteBuffer buf)
            where T : Serializable, new()
        {
            var value = new T();
            value.Decode(buf);
            return value;
        }

        public static Action<ByteBuffer, T> CreateEncodeFunc<T>()
        {
            var type = typeof(T);

            if (type == typeof(bool))
            {
                return (Action<ByteBuffer, T>)(Delegate)(new Action<ByteBuffer, bool>(
                        (ByteBuffer buf, bool x) => buf.WriteBool(x)));
            }

            if (type == typeof(byte))
            {
                return (Action<ByteBuffer, T>)(Delegate)(new Action<ByteBuffer, byte>(
                        (ByteBuffer buf, byte x) => buf.WriteLong((sbyte)x)));
            }

            if (type == typeof(short))
            {
                return (Action<ByteBuffer, T>)(Delegate)(new Action<ByteBuffer, short>(
                        (ByteBuffer buf, short x) => buf.WriteLong(x)));
            }

            if (type == typeof(int))
            {
                return (Action<ByteBuffer, T>)(Delegate)(new Action<ByteBuffer, int>(
                        (ByteBuffer buf, int x) => buf.WriteInt(x)));
            }

            if (type == typeof(long))
            {
                return (Action<ByteBuffer, T>)(Delegate)(new Action<ByteBuffer, long>(
                        (ByteBuffer buf, long x) => buf.WriteLong(x)));
            }

            if (type == typeof(string))
            {
                return (Action<ByteBuffer, T>)(Delegate)(new Action<ByteBuffer, string>(
                        (ByteBuffer buf, string x) => buf.WriteString(x)));
            }

            if (type == typeof(double))
            {
                return (Action<ByteBuffer, T>)(Delegate)(new Action<ByteBuffer, double>(
                        (ByteBuffer buf, double x) => buf.WriteDouble(x)));
            }

            if (type == typeof(float))
            {
                return (Action<ByteBuffer, T>)(Delegate)(new Action<ByteBuffer, float>(
                        (ByteBuffer buf, float x) => buf.WriteFloat(x)));
            }

            if (type == typeof(Binary))
            {
                return (Action<ByteBuffer, T>)(Delegate)(new Action<ByteBuffer, Binary>(
                        (ByteBuffer buf, Binary x) => buf.WriteBinary(x)));
            }

            if (typeof(Serializable).IsAssignableFrom(typeof(T)))
            {
                return (ByteBuffer buf, T x) => (x as Serializable).Encode(buf);
            }

            return null;
        }
    }

    public sealed class SerializeHelper<T> : SerializeHelper
    {
        public static Func<ByteBuffer, T> Decode { get; } = CreateDecodeFunc<T>();
        public static Action<ByteBuffer, T> Encode { get; } = CreateEncodeFunc<T>();
    }
}
